ci: label docker images and prune dangling images selectively

stickies-v opened the pull request:
Follow-up from https://github.com/bitcoin/bitcoin/pull/27777#discussion_r1210209382.

Labeling the docker images produced by the CI allows us/the user to apply batch operations to all images (including dangling ones) produced by the ci without affecting other, non-bitcoin-ci images. With labeling, we can safely always prune dangling bitcoin-ci-test images without checking for `RESTART_CI_DOCKER_BEFORE_RUN`, which we enable on our persistent runners.

stickies-v committed:
ci: add label to docker images

This allows us or the user to perform batch operations on all
images produced by the ci, e.g. to prune all dangling images,
without affecting non-ci images.

stickies-v committed:
ci: always prune all dangling bitcoin-ci-test images

Since all bitcoin-ci-test images are now labeled, we can always
prune all dangling images, regardless of whether we are in
RESTART_CI_DOCKER_BEFORE_RUN.

To be safe, still prune all images if RESTART_CI_DOCKER_BEFORE_RUN
in case the filtering doesn't work, or if images were created on
an earlier version that did not assign labels.

Sjors commented:
Concept ACK. Haven't tested.

It would also be nice to have an incantation (e.g. in the help) to kill dangling images after CI failure / interruption (if you don't plan on running it again).

stickies-v commented:
Rebased to address merge conflict from https://github.com/bitcoin/bitcoin/pull/27844.

> It would also be nice to have an incantation (e.g. in the help) to kill dangling images after CI failure / interruption (if you don't plan on running it again).

Thanks, added [doc: ci: add instructions for pruning dangling images manually](https://github.com/bitcoin/bitcoin/pull/27793/commits/f9b5fe40cdaab04a1bf72699f1eb4be44ab24bbe)

fanquake commented:
Concept ACK

stickies-v commented:
Rebased to address merge conflict from https://github.com/bitcoin/bitcoin/pull/28138.

stickies-v commented:
Rebased to address merge conflict from https://github.com/bitcoin/bitcoin/pull/28161.

stickies-v commented on ci/README.md:
````diff
@@ -26,6 +26,13 @@ To run the test stage with a specific configuration,
 ```
 
+The ci automatically prunes all dangling docker images every time it runs. If you want to perform
+the cleanup manually (e.g. if you won't be running the ci again), you can run
+
+```sh
+docker image prune --force --filter "label=bitcoin-ci-test"
````
Yeah, would you prefer it like this?

````diff
diff --git a/ci/README.md b/ci/README.md
index ab5e6ce866..16310b7703 100644
--- a/ci/README.md
+++ b/ci/README.md
@@ -26,13 +26,19 @@ To run the test stage with a specific configuration,
 FILE_ENV="./ci/test/00_setup_env_arm.sh" ./ci/test_run_all.sh
 ```
 
-The ci automatically prunes all dangling docker images every time it runs. If you want to perform
-the cleanup manually (e.g. if you won't be running the ci again), you can run
+The ci automatically prunes all dangling (i.e. unused) docker images every time it runs. To do this
+manually, run
 
 ```sh
 docker image prune --force --filter "label=bitcoin-ci-test"
 ```
 
+To also remove the used images (e.g. if you're not planning on running the ci again), run
+
+```sh
+docker rmi -f $(docker images -q -a --filter "label=bitcoin-ci-test")
+```
+
 ### Configurations
 
 The test files (`FILE_ENV`) are constructed to test a wide range of
````

stickies-v commented:
Force pushed to update documentation [addressing this review comment](https://github.com/bitcoin/bitcoin/pull/27793#discussion_r1293632896).

stickies-v commented on ci/README.md:
````diff
@@ -26,6 +26,19 @@ To run the test stage with a specific configuration,
+docker image prune --force --filter "label=bitcoin-ci-test"
+```
+
+To also remove the used images (e.g. if you're not planning on running the ci again), run
````
Thanks Dropped the README.md commit, don't feel like spending more time on this. @Sjors feel free to follow up if you think it's important.

stickies-v commented:
Rebased to address merge conflict from #27976.

pablomartin4btc reviewed (commented):
Concept ACK.

Perhaps description needs to be updated regarding the user instructions (`ci/README.md`).

stickies-v commented:
> Perhaps description needs to be updated regarding the user instructions (`ci/README.md`).

Thanks, reverted that description.

stickies-v commented:
Closed, not an essential change and doesn't seem like there's sufficient interest.

maflcko reviewed (commented):
lgtm, after the nit

stickies-v commented on ci/test/04_install.sh:
```diff
@@ -26,9 +28,9 @@ if [ -z "$DANGER_RUN_CI_ON_HOST" ]; then
     echo "Restart docker before run to stop and clear all containers started with --rm"
     podman container stop --all  # Similar to "systemctl restart docker"
-    echo "Prune all dangling images"
-    docker image prune --force
```
Makes sense, especially the migration aspect which I didn't think about. Added a comment that it can probably be removed in the future, though.

stickies-v commented:
Force-pushed to address merge conflict from #28547 and incorporate [suggestion](https://github.com/bitcoin/bitcoin/pull/27793#discussion_r1357995285) from maflcko to keep the unfiltered pruning for now, as belt and suspenders approach.

fanquake reviewed (approved):
utACK e44c574650827f18e12ac0ba378c0a19d23a07b4
